src/approximate-qvm.lisp

```diff
@@ -0,0 +1,132 @@
+   (readout-povms
+    :initarg :readout-povms
+    :accessor readout-povms
+    :initform (make-hash-table :test 'eql)))
+  (:documentation "APPROX-QVM is a QVM that supports a noise model defined by T1, T2, and readout fidelities on each qubit. At each instruction, T1 noise and T2 noise is applied for the qubits involved in the instruction. Upon MEASURE, the readout noise is also applied for whichever qubit is being measured."))
+
+
+(defun (setf qubit-t1) (t1 qvm qubit)
+  "Save the kraus operators for the T1 provided on the specific qubit."
+  (let ((kraus-ops (generate-damping-kraus-ops t1 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t1-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-t2) (t2 qvm qubit)
+  "Save the kraus operators for the T2  provided on the specified qubit."
+  (let ((kraus-ops (generate-damping-dephasing-kraus-ops t2 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t2-ops qvm)) kraus-ops)))
+
+
+(defmethod (setf qubit-fro) (fro (qvm approx-qvm) qubit)
+  ;; Save the readout povms for the fro provided on the specified
+  ;; qubit. The fR0 is assumed to be the average readout fidelity (avg
+  ;; of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) =
+  ;; p(1|1) = fR0 in the readout povm assignment probabilities.
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
+
+
+(defmethod apply-all-kraus-operators ((qvm approx-qvm) (instr quil:gate-application) kraus-ops)
+  ;; Apply all the kraus operators for each type of noise
+  ;; source. KRAUS-OPS is a list of ( (t1-operators) (t2-operators)),
+  ;; where t1-operators is a list of t1 noise maps for each qubit in
+  ;; the instruction.  If the instruction is a 2q gate, we need to
+  ;; tensor the kraus operators for each qubit's error source.
+  (loop :for noise-operators in kraus-ops
+        :do (let ((simplified-kraus-ops  (reduce #'kraus-kron noise-operators)))
+              (when simplified-kraus-ops
+                (apply-kraus-ops qvm instr simplified-kraus-ops)))))
+
+
+(defmethod transition ((qvm approx-qvm) (instr quil:gate-application))
+  ;; For the current instruction in the program, apply the gate
+  ;; corresponding to that instruction, and then apply all the kraus
+  ;; operators for all the sources of noise following that
+  ;; instruction.
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar (lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (t1-ops (loop :for q in qubits :collect (gethash q (t1-ops qvm))))
+         (t2-ops (loop :for q in qubits :collect (gethash q (t2-ops qvm))))
+         (kraus-ops (list t1-ops t2-ops)))
```

Comment:
Sure, I could do that! Would that be better to do stylistically? I originally thought it would be better for each `apply-all-kraus-operators` to be specialized on each qvm class for readability. I thought each class should have its own method to clearly express "This is how the kraus operators are applied for each type of noisy qvm". 

In this case, since the `apply-all-kraus-operators` on `approx-qvm` is so short, is it better to just have that code exist in `transition`?

Reply:
Hmmm good question. I think having a separate function for applying the kraus ops is reasonable. I guess I just noticed that the difference between the methods for `channel-qvm` and `approx-qvm` is small now after the latest refactoring and wondered if they could be combined. It also struck me that the difference that remains isn't really in how the operators are *applied* so much as how they are extracted/combined/whatever, but that might just be arguing semantics. At the end of the day, it's stylistic question as you say, so whatever you think is the right way to slice it is fine by me.

I definitely think it makes sense to have something like the existing `apply-kraus-ops` since that needs to happen in several places and is non-trivial. In fact, we should probably be calling that from inside `noisy-qvm`'s `transition` method as well (assuming the implementations are identical, I didn't look too closely).

Reply:
Oh, I just realized after reading your other comment about `apply-kraus-ops` that `noisy-qvm` and `channel-qvm` are "siblings", and so calling `apply-kraus-ops` from `noisy-qvm`'s `transition` method probably wouldn't work (since it's specialized on `channel-qvm`).